File: libs/document-processor/src/registry/processors/steps/update-document-step-executor.ts
import { BlueNode } from '@blue-labs/language';
import type { BexEngine } from '@blue-labs/bex';
import {
  UpdateDocumentSchema,
  type UpdateDocument,
} from '@blue-repository/types/packages/coordination/schemas/UpdateDocument';

import type { JsonPatch } from '../../../model/shared/json-patch.js';
import { conversationBlueIds } from '../../../repository/semantic-repository.js';
import type { ContractProcessorContext } from '../../types.js';
import type {
  SequentialWorkflowStepExecutor,
  StepExecutionArgs,
} from '../workflow/step-runner.js';
import { BexFieldEvaluator } from './bex-field-evaluator.js';

type JsonPatchOperation = 'ADD' | 'REPLACE' | 'REMOVE';

type ChangeInput = Required<UpdateDocument>['changeset'][number] | BlueNode;

export class UpdateDocumentStepExecutor implements SequentialWorkflowStepExecutor {
  readonly supportedBlueIds = [
    conversationBlueIds['Conversation/Update Document'],
  ] as const;

  private readonly bexEvaluator: BexFieldEvaluator;

  constructor(bexEngine?: BexEngine) {
    this.bexEvaluator = new BexFieldEvaluator(bexEngine);
  }

  async execute(args: StepExecutionArgs): Promise<unknown> {
    const { context, stepNode } = args;
    const { blue } = context;

    if (
      !blue.isTypeOfBlueId(
        stepNode,
        conversationBlueIds['Conversation/Update Document'],
      )
    ) {
      return context.throwFatal('Update Document step payload is invalid');
    }

    let resolvedStepNode = stepNode;
    const changesetNode = resolvedStepNode.getProperties()?.changeset;
    if (
      changesetNode !== undefined &&
      context.measure('bex.fieldEvaluation.containsExpression', () =>
        this.bexEvaluator.containsExpression(changesetNode),
      )
    ) {
      resolvedStepNode = resolvedStepNode.clone();
      resolvedStepNode.addProperty(
        'changeset',
        this.bexEvaluator.evaluateNode(args, changesetNode),
      );
    }
    const changeset = context.measure('bex.compute.toPatches', () =>
      this.extractChanges(resolvedStepNode, context),
    );

    context.gasMeter().chargeUpdateDocumentBase(changeset.length);
    for (const change of changeset) {
      const patch = this.createPatch(change, context);
      await context.applyPatch(patch);
    }

    return undefined;
  }

  private extractChanges(
    stepNode: BlueNode,
    context: ContractProcessorContext,
  ): ChangeInput[] {
    const changesetNode = stepNode.getProperties()?.changeset;
    if (!changesetNode) {
      return [];
    }
    if (
      context.measure('bex.fieldEvaluation.containsExpression', () =>
        this.bexEvaluator.containsExpression(changesetNode),
      )
    ) {
      return context.throwFatal(
        'Update Document changeset still contains unevaluated BEX',
      );
    }
    const items = changesetNode.getItems();
    if (items !== undefined) {
      return [...items];
    }
    if (changesetNode.getValue() !== undefined) {
      return context.throwFatal(
        'Update Document changeset must evaluate to a list',
      );
    }
    const schemaOutput = context.blue.nodeToSchemaOutput<UpdateDocument>(
      stepNode,
      UpdateDocumentSchema,
    );
    return schemaOutput.changeset ?? [];
  }

  private createPatch(
    change: ChangeInput,
    context: ContractProcessorContext,
  ): JsonPatch {
    const op = this.normalizeOperation(this.changeField(change, 'op'), context);
    const path = this.normalizePath(this.changeField(change, 'path'), context);
    const absolutePath = context.resolvePointer(path);

    if (op === 'REMOVE') {
      return { op, path: absolutePath };
    }

    const val = this.changeValue(change);
    if (!val) {
      return context.throwFatal(`${op} Update Document operation missing val`);
    }
    return { op, path: absolutePath, val };
  }

  private changeField(change: ChangeInput, key: string): unknown {
    if (change instanceof BlueNode) {
      return change.getProperties()?.[key]?.getValue();
    }
    return change[key as keyof ChangeInput];
  }

  private changeValue(change: ChangeInput): BlueNode | undefined {
    if (change instanceof BlueNode) {
      return change.getProperties()?.val;
    }
    return change.val;
  }

  private normalizeOperation(
    rawOp: unknown,
    context: ContractProcessorContext,
  ): JsonPatchOperation {
    const text = typeof rawOp === 'string' ? rawOp : undefined;
    const upper = (text ?? 'REPLACE').toUpperCase();

    if (upper === 'ADD' || upper === 'REPLACE' || upper === 'REMOVE') {
      return upper;
    }

    return context.throwFatal(
      `Unsupported Update Document operation "${text}"`,
    );
  }

  private normalizePath(
    rawPath: unknown,
    context: ContractProcessorContext,
  ): string {
    if (typeof rawPath !== 'string') {
      return context.throwFatal('Update Document changeset requires a path');
    }

    const trimmed = rawPath.trim();
    if (trimmed.length === 0) {
      return context.throwFatal('Update Document path cannot be empty');
    }

    return trimmed;
  }
}

File: references/java-sdk/src/main/java/blue/language/processor/workflow/steps/UpdateDocumentStepExecutor.java
package blue.language.processor.workflow.steps;

import blue.language.model.Node;
import blue.language.processor.model.JsonPatch;
import blue.language.processor.script.QuickJSEvaluator;
import blue.language.processor.script.QuickJsExpressionUtils;
import blue.language.processor.workflow.StepExecutionArgs;
import blue.language.processor.workflow.WorkflowStepExecutor;

import java.util.Collections;
import java.util.List;
import java.util.Locale;
import java.util.LinkedHashSet;
import java.util.Set;

public class UpdateDocumentStepExecutor implements WorkflowStepExecutor {

    private static final Set<String> SUPPORTED_BLUE_IDS =
            Collections.unmodifiableSet(new LinkedHashSet<String>(java.util.Arrays.asList(
                    "Conversation/Update Document",
                    "UpdateDocument"
            )));

    private final QuickJSEvaluator evaluator;

    public UpdateDocumentStepExecutor() {
        this(new QuickJSEvaluator());
    }

    public UpdateDocumentStepExecutor(QuickJSEvaluator evaluator) {
        this.evaluator = evaluator;
    }

    @Override
    public Set<String> supportedBlueIds() {
        return SUPPORTED_BLUE_IDS;
    }

    @Override
    public Object execute(StepExecutionArgs args) {
        if (!isValidStepNode(args.stepNode(), args)) {
            args.context().throwFatal("Update Document step payload is invalid");
            return WorkflowStepExecutor.NO_RESULT;
        }
        Node stepNode = QuickJsExpressionUtils.resolveExpressions(
                args.stepNode(),
                evaluator,
                QuickJSStepBindings.create(args),
                args.context(),
                QuickJsExpressionUtils.createPathPredicate(
                        java.util.Arrays.asList("/changeset", "/changeset/**"),
                        null),
                null);
        List<Node> changes = readChangeset(stepNode);
        args.context().chargeUpdateDocumentBase(changes.size());
        for (Node change : changes) {
            applyChange(args, change);
        }
        return WorkflowStepExecutor.NO_RESULT;
    }

    private List<Node> readChangeset(Node stepNode) {
        if (stepNode == null || stepNode.getProperties() == null) {
            return Collections.emptyList();
        }
        Node changesetNode = stepNode.getProperties().get("changeset");
        if (changesetNode == null || changesetNode.getItems() == null) {
            return Collections.emptyList();
        }
        return changesetNode.getItems();
    }

    private void applyChange(StepExecutionArgs args, Node change) {
        if (change == null || change.getProperties() == null) {
            args.context().throwFatal("Update Document changeset entry must be an object");
            return;
        }
        String op = readString(change, "op");
        if (op == null || op.trim().isEmpty()) {
            op = "REPLACE";
        }
        String normalizedOp = op.trim().toUpperCase(Locale.ROOT);
        String path = readPath(change);
        if (path == null || path.trim().isEmpty()) {
            args.context().throwFatal("Update Document changeset requires a non-empty path");
            return;
        }
        String absolutePath = args.context().resolvePointer(path);

        if ("REMOVE".equals(normalizedOp)) {
            args.context().applyPatch(JsonPatch.remove(absolutePath));
            return;
        }

        Node valueNode = change.getProperties().get("val");
        if (valueNode == null) {
            args.context().throwFatal("Update Document " + normalizedOp + " operation requires val");
            return;
        }
        if ("ADD".equals(normalizedOp)) {
            args.context().applyPatch(JsonPatch.add(absolutePath, valueNode.clone()));
            return;
        }
        if ("REPLACE".equals(normalizedOp)) {
            args.context().applyPatch(JsonPatch.replace(absolutePath, valueNode.clone()));
            return;
        }
        args.context().throwFatal("Unsupported Update Document operation \"" + op + "\"");
    }

    private String readPath(Node node) {
        if (node == null || node.getProperties() == null) {
            return null;
        }
        Node pathNode = node.getProperties().get("path");
        if (pathNode == null || pathNode.getValue() == null) {
            return null;
        }
        Object value = pathNode.getValue();
        if (!(value instanceof String)) {
            return null;
        }
        return (String) value;
    }

    private String readString(Node node, String key) {
        if (node.getProperties() == null) {
            return null;
        }
        Node valueNode = node.getProperties().get(key);
        if (valueNode == null || valueNode.getValue() == null) {
            return null;
        }
        return String.valueOf(valueNode.getValue());
    }

    private boolean isValidStepNode(Node stepNode, StepExecutionArgs args) {
        return WorkflowStepTypeSupport.isSupportedStepType(
                stepNode,
                SUPPORTED_BLUE_IDS,
                args != null && args.context() != null ? args.context().nodeProvider() : null);
    }
}
